### PR TITLE
Threshold line for HSR #727

### LINKED ISSUE #727: Threshold line for HSR
Add line that represents threshold by analogy with anomaly bounds

related issue in app https://github.com/hastic/hastic-grafana-app/issues/315

### PR DESCRIPTION
**Pull request by amper43**
fixes #727 

## Changes
- threshold detector's `process_data()` returns `ProcessingResult`
- add lower / upper bound depend on threshold condition

### CODE REVIEW

File: analytics/analytics/detectors/threshold_detector.py [hunk 1]
```diff
@@ -89,3 +89,23 @@ def concat_detection_results(self, detections: List[DetectionResult]) -> Detecti
             result.cache = detection.cache
         result.segments = utils.merge_intersecting_segments(result.segments, time_step)
         return result
+
+    def process_data(self, dataframe: pd.DataFrame, cache: ModelCache) -> ProcessingResult:
```
Comment (jonyrock) [inline analytics/analytics/detectors/threshold_detector.py]: I need a doc for this method, can't understand the purpose of if from the name

File: server/src/controllers/analytics_controller.ts [hunk 2]
```diff
@@ -30,6 +30,12 @@ type TableTimeSeries = { values: [number, number][], columns: string[] };
 type TimeRange = { from: number, to: number };
 export type TaskResolver = (taskResult: TaskResult) => void;
 
+type hsrResult = {
```
Comment (jonyrock) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
type HSRResult = {
```

File: server/src/controllers/analytics_controller.ts [hunk 3]
```diff
@@ -631,10 +637,14 @@ export async function getHSR(
   try {
     const grafanaUrl = getGrafanaUrl(analyticUnit.grafanaUrl);
     const data = await queryByMetric(analyticUnit.metric, grafanaUrl, from, to, HASTIC_API_KEY);
+    let resultSeries: hsrResult = {
```
Comment (jonyrock) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
    let hsrResult: HSRResult = {
```

File: analytics/analytics/analytic_types/detector_typing.py [hunk 4]
```diff
@@ -32,17 +32,6 @@ def to_json(self):
...
-    def __init__(
-        self,
-        data: Optional[TimeSeries] = None
-    ):
-        if data is None:
-            data = []
-        self.data = data
-
-@utils.meta.JSONClass
-class AnomalyProcessingResult():
```
Comment (jonyrock) [inline analytics/analytics/analytic_types/detector_typing.py]: please explain this change

File: analytics/analytics/detectors/threshold_detector.py [hunk 5]
```diff
@@ -89,3 +89,23 @@ def concat_detection_results(self, detections: List[DetectionResult]) -> Detecti
...
+        data = dataframe['value']
+        value = cache['value']
+        condition = cache['condition']
+
+        if condition == 'NO_DATA':
+            return ProcessingResult()
+
+        data.values[:]  = value
```
Comment (jonyrock) [inline analytics/analytics/detectors/threshold_detector.py]:
```suggestion
        data.values[:] = value
```

File: analytics/analytics/detectors/threshold_detector.py [hunk 6]
```diff
@@ -89,3 +89,23 @@ def concat_detection_results(self, detections: List[DetectionResult]) -> Detecti
             result.cache = detection.cache
         result.segments = utils.merge_intersecting_segments(result.segments, time_step)
         return result
+
+    def process_data(self, dataframe: pd.DataFrame, cache: ModelCache) -> ProcessingResult:
+        data = dataframe['value']
+        value = cache['value']
+        condition = cache['condition']
```
Comment (jonyrock) [inline analytics/analytics/detectors/threshold_detector.py]: Please resolve the merge conflict, there is a new function for thing like this line

File: analytics/analytics/detectors/threshold_detector.py [hunk 7]
```diff
@@ -89,3 +89,23 @@ def concat_detection_results(self, detections: List[DetectionResult]) -> Detecti
...
+        timestamps = utils.convert_series_to_timestamp_list(dataframe.timestamp)
+        result_series = list(zip(timestamps, data.values.tolist()))
+
+        if condition in ['>', '>=', '=']:
+            return ProcessingResult(upper_bound=result_series)
+
+        if condition in ['<', '<=']:
+            return ProcessingResult(lower_bound=result_series)
+
+        raise ValueError(f'{condition} condition not supported')
```
Comment (jonyrock) [inline analytics/analytics/detectors/threshold_detector.py]:
```suggestion
        raise ValueError(f'{condition} condition not supported')

```

File: analytics/analytics/detectors/anomaly_detector.py [hunk 8]
```diff
@@ -223,7 +223,7 @@ def process_data(self, dataframe: pd.DataFrame, cache: ModelCache) -> AnomalyPro
         result_bounds = {}
         for bound_type, bound_data in bounds.items():
             result_bounds[bound_type] = list(zip(timestamps, bound_data.values.tolist()))
-        result = AnomalyProcessingResult(lower_bound=result_bounds.get(Bound.LOWER.value), upper_bound=result_bounds.get(Bound.UPPER.value))
+        result = ProcessingResult(lower_bound=result_bounds.get(Bound.LOWER.value), upper_bound=result_bounds.get(Bound.UPPER.value))
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]: Let's merge `master` and modify refactored code

File: analytics/analytics/detectors/threshold_detector.py [hunk 9]
```diff
@@ -89,3 +89,23 @@ def concat_detection_results(self, detections: List[DetectionResult]) -> Detecti
...
+
+        if condition == 'NO_DATA':
+            return ProcessingResult()
+
+        data.values[:]  = value
+        timestamps = utils.convert_series_to_timestamp_list(dataframe.timestamp)
+        result_series = list(zip(timestamps, data.values.tolist()))
+
+        if condition in ['>', '>=', '=']:
+            return ProcessingResult(upper_bound=result_series)
```
Comment (rozetko) [inline analytics/analytics/detectors/threshold_detector.py]: Let's add spaces before and after `=`

File: analytics/analytics/analytic_types/detector_typing.py (hunk 4 above)
Comment (amper43) [reply analytics/analytics/analytic_types/detector_typing.py]: AnomalyProcessingResult represented result interface. Now both treshold and anomaly detectors are childs of `ProcessingDetectors` and both have the same result interface. In the past only anomaly detector had this interface. In this PR result's class renames to `ProcessingResult` and starts use in threshold detector.

File: analytics/analytics/detectors/detector.py [hunk 10]
```diff
@@ -53,14 +53,28 @@ class ProcessingDetector(Detector):
...
     def process_data(self, data: TimeSeries, cache: Optional[ModelCache]) -> ProcessingResult:
+        '''
+        Data processing to receive additional time series that represents detector's settings
+        '''
         pass
 
     def concat_processing_results(self, processing_results: List[ProcessingResult]) -> Optional[ProcessingResult]:
+        '''
+        Concatenate sequential ProcessingResults that received via
+        split pervious dataset to chunks in analytic worker
```
Comment (rozetko) [inline analytics/analytics/detectors/detector.py]:
```suggestion
        splitting dataset to chunks in analytic worker
```

**Pull request merged**